### ISSUE TITLE
[ASCollectionView] Not calling reloadData with a supplementary node will cause a crash

### ISSUE THREAD

**Issue by smyrgl** [facebook/AsyncDisplayKit#1322]
Sample project: https://github.com/smyrgl/adk-test

I was finally able to nail down this crash when creating a sample project to isolate it.  If you don't call `reloadData` on the collection node it will crash, it looks like because the ASCollectionViewLayoutInspecting class is returning a section "count" that differs from the section count returned by the collection node datasource at this time.  This might be considered relatively minor since you can have them return from the same source potentially but I'm not sure the "desynchronization" here of these datasources is very safe.

```
2016-03-04 10:05:06.488 adk_test[44865:24652742] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'request for layout attributes for supplementary view UICollectionElementKindSectionHeader in section 0 when there are only 0 sections in the collection view'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000010149fe65 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x00000001031dfdeb objc_exception_throw + 48
	2   CoreFoundation                      0x000000010149fcca +[NSException raise:format:arguments:] + 106
	3   Foundation                          0x00000001018e84de -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:] + 198
	4   UIKit                               0x000000010256d0ee -[UICollectionViewData layoutAttributesForSupplementaryElementOfKind:atIndexPath:] + 311
	5   UIKit                               0x0000000102520e72 -[UICollectionView _dequeueReusableViewOfKind:withIdentifier:forIndexPath:viewCategory:] + 354
	6   UIKit                               0x0000000102521a8f -[UICollectionView dequeueReusableSupplementaryViewOfKind:withReuseIdentifier:forIndexPath:] + 288
	7   AsyncDisplayKit                     0x00000001007d98ae -[ASCollectionView collectionView:viewForSupplementaryElementOfKind:atIndexPath:] + 174
	8   UIKit                               0x0000000102511fb1 -[UICollectionView _createPreparedSupplementaryViewForElementOfKind:atIndexPath:withLayoutAttributes:applyAttributes:] + 679
	9   UIKit                               0x0000000102513b18 -[UICollectionView _updateVisibleCellsNow:] + 4487
	10  UIKit                               0x000000010251823b -[UICollectionView layoutSubviews] + 247
	11  AsyncDisplayKit                     0x00000001007db430 -[ASCollectionView layoutSubviews] + 784
	12  UIKit                               0x0000000101d734a3 -[UIView(CALayerDelegate) layoutSublayersOfLayer:] + 703
	13  QuartzCore                          0x000000010123259a -[CALayer layoutSublayers] + 146
	14  AsyncDisplayKit                     0x00000001007a8aa7 -[_ASDisplayLayer layoutSublayers] + 55
	15  QuartzCore                          0x0000000101226e70 _ZN2CA5Layer16layout_if_neededEPNS_11TransactionE + 366
	16  QuartzCore                          0x0000000101226cee _ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE + 24
	17  QuartzCore                          0x000000010121b475 _ZN2CA7Context18commit_transactionEPNS_11TransactionE + 277
	18  QuartzCore                          0x0000000101248c0a _ZN2CA11Transaction6commitEv + 486
	19  QuartzCore                          0x000000010124937c _ZN2CA11Transaction17observer_callbackEP19__CFRunLoopObservermPv + 92
	20  CoreFoundation                      0x00000001013cb367 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 23
	21  CoreFoundation                      0x00000001013cb2d7 __CFRunLoopDoObservers + 391
	22  CoreFoundation                      0x00000001013c084c CFRunLoopRunSpecific + 524
	23  UIKit                               0x0000000101cb77cd -[UIApplication _run] + 402
	24  UIKit                               0x0000000101cbc610 UIApplicationMain + 171
	25  adk_test                            0x0000000100604d4d main + 109
	26  libdyld.dylib                       0x0000000103d2092d start + 1
	27  ???                                 0x0000000000000001 0x0 + 1
```

**Comment by levi** [comment]
@smyrgl update on this. Looks to be related to ASViewController. `reloadData` should be called automatically when the collection view's `layoutSubviews` gets called for the first time. Continuing my investigation, but if you need to work around at the moment use a `UIViewController`.

**Comment by smyrgl** [comment]
@levi A simpler solution (since ASViewController and UIViewController have different initializers) is just to do a `reloadData` call in `viewDidLoad` AFTER I register the supplementary views.  That seems to work fine for now.

**Comment by levi** [comment]
@smyrgl I've taken another stab at your example project tonight and noticed that the project runs fine if I switch the layout to use a vanilla `UICollectionViewFlowLayout` instance. I'm going to take a closer look at your layout implementation, but was curious if you've noticed anything on your end regarding the custom flow layout subclass?

**Comment by maicki** [comment]
@smyrgl Can you verify if this is still an issue. Just ran the sample project and it seems to be fixed with latest master or 1.9.73

**Comment by appleguy** [comment]
@smyrgl: It would be great if you could try out @erichoracek's awesome contribution and confirm that this works for you!
